File: 01-machine/02-threads-and-syscalls/java/SyscallCost.java
// Layer 1 - What a syscall actually costs, Java version.
// FileInputStream.read() on /dev/zero eventually calls read(2), but there
// are several layers between your Java call and the kernel: a JNI native
// method call, a transition out of the JVM's managed execution, argument
// marshalling -- more machinery than C++'s direct call, which is part of
// why this measures higher overhead than C++ or Go despite hitting the
// exact same kernel function.
import java.io.FileInputStream;
import java.io.IOException;

public class SyscallCost {
    static final long N = 500_000;

    static double benchSyscall() throws IOException {
        FileInputStream in = new FileInputStream("/dev/zero");
        byte[] buf = new byte[1];
        long start = System.nanoTime();
        for (long i = 0; i < N; i++) {
            in.read(buf);
        }
        long elapsed = System.nanoTime() - start;
        in.close();
        return elapsed / 1e9;
    }

    static double benchPureJava() {
        long total = 0;
        long start = System.nanoTime();
        for (long i = 0; i < N; i++) {
            total += i & 0xFF;
        }
        long elapsed = System.nanoTime() - start;
        if (total == -1) System.out.println(""); // keep the loop from being optimized away
        return elapsed / 1e9;
    }

    public static void main(String[] args) throws IOException {
        double tSys = benchSyscall();
        double tPure = benchPureJava();
        System.out.printf("N=%d%n", N);
        System.out.printf("read(/dev/zero) x%d:  %6.3fs  (%6.1f ns/call)%n", N, tSys, tSys / N * 1e9);
        System.out.printf("pure Java loop:       %6.3fs  (%6.1f ns/iter)%n", tPure, tPure / N * 1e9);
        System.out.printf("syscall is %.1fx the cost of an equivalent pure-Java step%n", tSys / tPure);
    }
}

File: 01-machine/02-threads-and-syscalls/nodejs/syscall_cost.js
// Layer 1 - What a syscall actually costs.
// Same read(/dev/zero) vs pure-loop comparison as the Python version.

const fs = require("fs");

const N = 500_000;

function benchSyscall() {
  const fd = fs.openSync("/dev/zero", "r");
  const buf = Buffer.alloc(1);
  const start = process.hrtime.bigint();
  for (let i = 0; i < N; i++) {
    fs.readSync(fd, buf, 0, 1, null);
  }
  const elapsed = Number(process.hrtime.bigint() - start) / 1e9;
  fs.closeSync(fd);
  return elapsed;
}

function benchPureJs() {
  let total = 0;
  const start = process.hrtime.bigint();
  for (let i = 0; i < N; i++) {
    total += i & 0xff;
  }
  const elapsed = Number(process.hrtime.bigint() - start) / 1e9;
  return elapsed;
}

const tSys = benchSyscall();
const tPure = benchPureJs();
console.log(`N=${N.toLocaleString()}`);
console.log(`read(/dev/zero) x${N}:  ${tSys.toFixed(3)}s  (${((tSys / N) * 1e9).toFixed(1)} ns/call)`);
console.log(`pure JS loop:           ${tPure.toFixed(3)}s  (${((tPure / N) * 1e9).toFixed(1)} ns/iter)`);
console.log(`syscall is ${(tSys / tPure).toFixed(1)}x the cost of an equivalent pure-JS step`);
